cpu/stm32f0/periph/gpio.c

```diff
@@ -24,315 +24,126 @@
     }
-
-    return 0; /* all OK */
+    return 0;
 }
 
-int gpio_init_int(gpio_t dev, gpio_pp_t pullup, gpio_flank_t flank, gpio_cb_t cb, void *arg)
+int gpio_init_int(gpio_t pin, gpio_pp_t pullup, gpio_flank_t flank, gpio_cb_t cb, void *arg)
 {
-    int res;
-    uint8_t pin;
-
-    if (dev >= GPIO_NUMOF) {
-        return -1;
-    }
+    int pin_num = _pin_num(pin);
+    int port_num = _port_num(pin);
 
-    pin = gpio_pin_map[dev];
-
-    /* configure pin as input */
-    res = gpio_init(dev, GPIO_DIR_IN, pullup);
-    if (res < 0) {
-        return res;
-    }
+    /* set callback */
+    exti_ctx[pin_num].cb = cb;
+    exti_ctx[pin_num].arg = arg;
 
-    /* set interrupt priority (its the same for all EXTI interrupts) */
-    NVIC_SetPriority(EXTI0_1_IRQn, GPIO_IRQ_PRIO);
-    NVIC_SetPriority(EXTI2_3_IRQn, GPIO_IRQ_PRIO);
-    NVIC_SetPriority(EXTI4_15_IRQn, GPIO_IRQ_PRIO);
+    /* enable the SYSCFG clock */
+    RCC->APB2ENR |= RCC_APB2ENR_SYSCFGEN;
 
     /* enable clock of the SYSCFG module for EXTI configuration */
     RCC->APB2ENR |= RCC_APB2ENR_SYSCFGCOMPEN;
 
-    /* read pin number, set EXIT channel and enable global interrupt for EXTI channel */
-    switch (dev) {
-#ifdef GPIO_0_EN
-        case GPIO_0:
-            GPIO_0_EXTI_CFG();
-            break;
-#endif
-#ifdef GPIO_1_EN
-        case GPIO_1:
-            GPIO_1_EXTI_CFG();
-            break;
-#endif
-#ifdef GPIO_2_EN
-        case GPIO_2:
-            GPIO_2_EXTI_CFG();
-            break;
-#endif
-#ifdef GPIO_3_EN
-        case GPIO_3:
-            GPIO_3_EXTI_CFG();
-            break;
-#endif
-#ifdef GPIO_4_EN
-        case GPIO_4:
-            GPIO_4_EXTI_CFG();
-            break;
-#endif
-#ifdef GPIO_5_EN
-        case GPIO_5:
-            GPIO_5_EXTI_CFG();
-            break;
-#endif
-#ifdef GPIO_6_EN
-        case GPIO_6:
-            GPIO_6_EXTI_CFG();
-            break;
-#endif
-#ifdef GPIO_7_EN
-        case GPIO_7:
-            GPIO_7_EXTI_CFG();
-            break;
-#endif
-#ifdef GPIO_8_EN
-        case GPIO_8:
-            GPIO_8_EXTI_CFG();
-            break;
-#endif
-#ifdef GPIO_9_EN
-        case GPIO_9:
-            GPIO_9_EXTI_CFG();
-            break;
-#endif
-#ifdef GPIO_10_EN
-        case GPIO_10:
-            GPIO_10_EXTI_CFG();
-            break;
-#endif
-#ifdef GPIO_11_EN
-        case GPIO_11:
-            GPIO_11_EXTI_CFG();
-            break;
-#endif
+
+    /* initialize pin as input */
+    gpio_init(pin, GPIO_DIR_IN, pullup);
+
+    /* enable global pin interrupt */
+    if (pin_num < 5) {
+        NVIC_EnableIRQ(EXTI0_IRQn + pin_num);
+    }
+    else if (pin_num < 10) {
+        NVIC_EnableIRQ(EXTI9_5_IRQn);
+    }
+    else {
+        NVIC_EnableIRQ(EXTI15_10_IRQn);
     }
-    NVIC_EnableIRQ(gpio_irq_map[dev]);
 
-    /* set callback */
-    gpio_config[dev].cb = cb;
-    gpio_config[dev].arg = arg;
+
 
     /* configure the event that triggers an interrupt */
     switch (flank) {
```

Comment:
This you could also optimize slightly.
You override the value of the `gpio_flank_t` type in the `periph_cpu.h` to these values:
```c
#define HAVE_GPIO_FLANK_T
typedef enum {
    GPIO_FALLING = 1,       /**< emit interrupt on falling flank */
    GPIO_RISING = 2,        /**< emit interrupt on rising flank */
    GPIO_BOTH = 3           /**< emit interrupt on both flanks */
} gpio_flank_t;
```

then drop the switch statement and write this:
```c
EXTI->RTSR &= ~(1 << pin_num);
EXTI->RTSR |= (((flank >> 1) & 0x1) << pin_num);
EXTI->FTSR &= ~(1 << pin_num);
EXTI->FTSR |= ((flank & 0x1) << pin_num);
```